src/psyclone/psyir/symbols/datasymbol.py

```diff
@@ -0,0 +1,539 @@
+    :param constant_value: sets a fixed known value for this DataSymbol. If \
+        the value is None (the default) then this symbol is not a constant. \
+        The datatype of the constant value must be compatible with the \
+        datatype of the symbol.
+    :type constant_value: int, str or bool
+    :param precision: the amount of storage required by the datatype (bytes) \
+            or a reference to a Symbol holding the type information \
+            or a label identifying a default precision.
+    :type precision: int or :py:class:`psyclone.psyir.symbol.DataSymbol` or \
+                     :py:class:`psyclone.psyir.symbols.DataSymbol.Precision`
+
+    :raises NotImplementedError: provided parameters are not supported yet.
+    :raises TypeError: provided parameters have invalid error type.
+    :raises ValueError: provided parameters contain invalid values.
+
+    '''
+    ## Tuple with the valid datatypes.
+    valid_data_types = ('real',  # Floating point
+                        'integer',
+                        'character',
+                        'boolean',
+                        'deferred')  # Type of this symbol not yet determined
+
+    ## Mapping from supported data types for constant values to
+    #  internal Python types
+    mapping = {'integer': int,
+               'character': str,
+               'boolean': bool,
+               'real': float}
+
+    class Precision(Enum):
+        '''
+        Enumeration for the different types of 'default' precision that may
+        be specified for a Symbol.
+        '''
+        SINGLE = 1
+        DOUBLE = 2
+
+    def __init__(self, name, datatype, shape=None, constant_value=None,
+                 interface=None, precision=None):
+
+        super(DataSymbol, self).__init__(name)
+
+        self._datatype = None
+        self.datatype = datatype
+
+        # Check that the supplied 'precision' is valid
+        if precision is not None:
+            if datatype.lower() not in ["real", "integer"]:
+                raise ValueError(
+                    "A DataSymbol of {0} type cannot have an associated "
+                    "precision".format(datatype.lower()))
+            if not isinstance(precision,
+                              (int, DataSymbol.Precision, DataSymbol)):
+                raise TypeError(
+                    "DataSymbol precision must be one of integer, "
+                    "DataSymbol.Precision or DataSymbol but got '{0}'"
+                    "".format(type(precision).__name__))
+            if isinstance(precision, int) and precision <= 0:
+                raise ValueError(
+                    "The precision of a DataSymbol when specified as an "
+                    "integer number of bytes must be > 0 but got {0}"
+                    "".format(precision))
+            if (isinstance(precision, DataSymbol) and
+                (precision.datatype not in ["integer", "deferred"]
```

Comment:
pylint complains about the indentation here.

Reply:
fixed